### ISSUE TITLE
Safari on OS X uses the wrong picture

### ISSUE THREAD

**Issue by alekseyg** [RealFaviconGenerator/realfavicongenerator#90]
On all the sites I've tried the RFG on, Safari on OS X uses a higher resolution picture than it should. It should take 16x16 on non-retina and 32x32 on Retina. It ends up using another. I noticed this when I tweaked the 16x16 and 32x32 icons to have less detail and thicker lines on one project. While Chrome and Firefox used the correct pictures, Safari didn't. Removing the HTML code (and probably making Safari use the .ico file) fixed the issue. Note: I replaced the PNG icons as well as the images inside the .ico file.

**Comment by alekseyg** [comment]
After testing different combinations, it seems that OS X Safari (using Mavericks and Safari 7.1) grabs whatever first icon meta tag it encounters in the HTML rather than using the correct size.

Edit: Observation 2: Adding the Favicon.ico into the HTML doesn't help Safari either if the PNG icons are present. This is such an odd bug. What is Safari's Issue?

**Comment by phbernard** [comment]
Hi Aleksey, thank you for reporting.

I don't have a Mac at my disposal, so I admit testing this platform is a bit complicated.

Could you run the four <a href="http://realfavicongenerator.net/favicon_compatibility_test">compatibility tests</a>? At least we will know exactly which icons are used by Safari in various situations (declaring `favicon.ico` in the HTML was a good move).

In your comment, you mention Safari takes the first declared PNG icon, is that right? Does that mean that moving the 16x16 delcaration first fixes the issue?

```
<link rel="icon" type="image/png" href="/favicon-16x16.png" sizes="16x16">
<link rel="icon" type="image/png" href="/favicon-192x192.png" sizes="192x192">
<link rel="icon" type="image/png" href="/favicon-160x160.png" sizes="160x160">
<link rel="icon" type="image/png" href="/favicon-96x96.png" sizes="96x96">
<link rel="icon" type="image/png" href="/favicon-32x32.png" sizes="32x32">
```

**Comment by alekseyg** [comment]
I ran the compatibility tests and in every instance I got the 192x192 icon except for the home screen/desktop test, in which case no icon appeared (this seems to be the case with any website I've tried).

Moving the 16x16 up icon doesn't fix the issue entirely. The problem is that a good portion of Mac users are using the new retina MacBook Pro's, which would need to use the 32x32 icon instead of 16x16.

Also, Safari includes a "Reading List" in addition to Bookmarks, which are basically bookmarks that can be accessed offline and continuously scrolled through. On non-retina Macs (or displays hooked up to retina Macs), Safari uses the 32x32 icon in the sidebar of the Reading List, so the non-retina Safari needs both 16x16 and 32x32.

One curious bug in Safari 7.x on a retina Mac is that it will not use the 64x64 icon in the Reading List even if it is using the .ico file which contains the 64x64 size or if is given a PNG icon of 64x64+, so the icon will be scaled up from 32x32 or 16x16, whichever is available, and always looks blurry. In the case of .ico, it uses the 32x32 size included, and in the case of a larger PNG icon, it downscales it for the URL bar to 32x32, and then upscales it from that for the Reading List.

In conclusion, non-retina requires both 16x16 and 32x32 while retina requires just 32x32. The only way I've seen Safari (on a non-retina display) use several sizes is through the .ico file when no PNG icons are declared at all, which can cause problems for other platforms. So unless a way to work around the bug completely is found, ~~perhaps the safest compromise would be to put the 32x32 PNG icon at the top and let non-retina Macs scale it down to 16x16.~~

Edit: After testing more (and reading what others have written before me), I see that Firefox and Chrome simply choose the last PNG while Safari and IE 11 choose the first, so the above suggestion is problematic as well.

**Comment by alekseyg** [comment]
Further testing reveals that IE 11 on Windows 8.1 has the same behavior. Now that I look at the [Compatibility Table](http://realfavicongenerator.net/favicon_compatibility), I see that this behavior is documented, but it is green, as though it's the expected behavior, when it probably shouldn't be. You don't need – nor want in many cases – the browser to download the largest icon for the tab/address bar for two reasons: 1) it takes more bandwidth and 2) the small icons are often tailored for simpler and sharper details at the tiny size.

**Comment by alekseyg** [comment]
I don't know if this is an optimal solution for the RFG, but for myself, I'm going to try to figure out a way to have desktop browsers use the .ico format, as current browsers seem to know what to do with it better than with PNG versions and the sizes attribute, while still working with Android and the large PNG image it requires. If I come up with something decent, I will post it here. I'm going to take precision over performance in this case (browsers cache the favicons more "permanently" than other resources, anyway). It's a shame favicons and browsers are such a mess.

**Comment by alekseyg** [comment]
OK, so here's the solution I came up with that works on every device (although the performance probably isn't great). I've included the .ico before all PNG files and then once more after. Here is an example:

Live site:
http://icon-test.hypernovastudio.com

GitHub Pages repo:
https://github.com/Hypernova-Design/icon-test

HTML for icons in particular:
https://github.com/Hypernova-Design/icon-test/blob/master/_includes/icons.html

Some notes:
1) I realize my .ico file is a bit of overkill as it contains the following sizes: 16, 24, 32, 48, 64, 128, 256. Windows will use 16 normally, but if you put it on the desktop, Windows 8 and 8.1 will use all of the following: 16, 24, 32, 48. If you are using Windows on a HiDPI device, it'll also use 64 and 128. Unless you enlarge your icons, it probably will not ever actually need 256. My testing wasn't exhaustive, but the only unused resolution was 256 in my testing. That doesn't necessarily mean you will still need all the resolutions, as they add considerable weight to the file. I was curious as to which resolutions are used on which devices and configurations, however.

2) Opera 12.15 seems to favor the 192x192 PNG over the 160x160 on Mac, retina or otherwise, so I wonder if the 160x160 PNG may be redundant (perhaps other platforms behave differently).

3) I believe the 16x16 and 32x32 PNGs are redundant with this setup as they are downloaded by some browsers, but never actually used by any browser/device I tested.

4) The (only) difference between this approach and totally omitting the HTML for the PNG favicons is that Android correctly chooses the 192x192 PNG icon over the apple-touch-icon in all places but the speed dial screen and bookmarks (same behavior as in the RFG compatibility tests).

5) Precision wise, this approach works better for desktop browsers, but still has the same shortcomings on Android (vanilla Android on a Nexus device using Chrome to be exact). ~~I will try working further on Android and see what I can do with it.~~ I can't seem to find any optimal solutions to this problem. Either we can use the .ico for the tab image and the apple-touch-icon for the home screen, or the 192 PNG file for both. The apple-touch-icon is still used in bookmarks and speed dial no matter what you do.

Update: Locally, everything worked fine. Once I uploaded everything to GitHub pages, everything went random again. How lovely!

**Comment by alekseyg** [comment]
OK, I think I have found to most optimal solution I could. The site is live at the same address, but the updated code is in the gh-pages branch of my repo.

Here's the updated code:
https://github.com/Hypernova-Design/icon-test/blob/gh-pages/_includes/icons.html

The relevant part:

``` html
<link rel="icon" type="image/x-icon" href="/favicon.ico" sizes="16x16 32x32 48x48 256x256">
<link rel="icon" type="image/png" href="/favicon-192x192.png" sizes="192x192">
<link rel="icon" type="image/png" href="/favicon-96x96.png" sizes="96x96">
<link rel="icon" type="image/x-icon" href="/favicon.ico" sizes="16x16 32x32 48x48 256x256">
```

The breakdown:
1. The most optimal resolutions to include in the .ico file for basically everything from Mac to Windows XP to Windows 8 is 16x16, 32x32, and 48x48. These are the sizes at which the icons are most often rendered. Though they may be rendered in other sizes, including them can mess up the common cases, and are therefore not worth it (also they add to the file size). The 256x256 size is used by Windows in HiDPI mode, and is compressed usually, so it doesn't add much to the file size. Source: http://msdn.microsoft.com/en-us/library/dn742485.aspx
2. Firefox loads all icons, so it's best to have a minimum in the HTML. Firefox and Chrome use the last one, while Safari and IE 11 use the first one. IE 10- has its own thing, but this still works fine. This way, the optimal size is almost always used on every desktop OS I tested. Compatibility table below.
3. Google TV should work fine. I don't have one to test with though. Android still works exactly the same as it did before with the original RFG code.

Compatibility table:

| Device | OS | Browser | Tab/URL | Bookmark | Home screen | Other |
| --- | --- | --- | --- | --- | --- | --- |
| Mac | OS X 10.9 | Safari 7.1 | :+1: 16x16 | :+1: 16x16 | N/A | Reading List: :+1: 32x32 |
| Mac | OS X 10.9 | Chrome 38 | :+1: 16x16 | :+1: 16x16 | N/A |  |
| Mac | OS X 10.9 | Firefox 32 | :+1: 16x16 | :+1: 16x16 | N/A |  |
| Mac | OS X 10.9 | Opera 12.15 | :-1: random | :-1: random | N/A | Speed Dial: :+1: 192x192 |
| Mac (retina) | OS X 10.9 | Safari 7.1 | :+1: 32x32 | :+1: 32x32 | N/A | Reading List: :-1: 32x32 (expected higher; Safari bug) |
| Mac (retina) | OS X 10.9 | Chrome 38 | :+1: 32x32 | :+1: 32x32 | N/A |  |
| Mac (retina) | OS X 10.9 | Firefox 32 | :+1: 32x32 | :-1: 16x16 (expected 32x32; Firefox bug) | N/A |  |
| Mac (retina) | OS X 10.9 | Opera 12.15 | :-1: random | :-1: random | N/A | Speed Dial: :+1: 192x192 |
| PC | Windows 8.1 | IE 11 | :+1: 16x16 | :+1: 16x16 | :+1: 48x48 | Various places: :+1: 32x32 |
| PC | Windows 8.1 (HiDPI) | IE 11 | :+1: 32x32 | :+1: 32x32 | :+1: 256x256 | Various places: :+1: 48x48 |
| PC | Windows 8.1 (HiDPI) | Chrome 38 | :+1: 32x32 | :+1: 32x32 | :+1: 256x256 | Various places: :+1: 48x48 |
| PC | Windows 8.1 (HiDPI) | Firefox 32 | :+1: 32x32 | :-1: 16x16 (expected 32x32; Firefox bug) | :-1: 16x16 (expected 256x256) |  |
| PC | Windows 8 | IE 10 | :+1: 16x16 | :+1: 16x16 | :+1: 48x48 | Various places: :+1: 32x32 |
| PC | Windows 7 | IE 11 | :+1: 16x16 | :+1: 16x16 | :+1: 48x48 | Various places: :+1: 32x32 |
| PC | Windows 7 | IE 10 | :+1: 16x16 | :+1: 16x16 | :+1: 48x48 | Various places: :+1: 32x32 |
| PC | Windows 7 | IE 9 | :+1: 16x16 | :+1: 16x16 | :+1: 48x48 | Various places: :+1: 32x32 |
| PC | Windows 7 | IE 8 | :-1: 16x16 (doesn't always show up) | :+1: 16x16 | :+1: 48x48 | Various places: :+1: 32x32 |
| PC | Windows XP–8.1 | Chrome 38 | :+1: 16x16 | :+1: 16x16 | :+1: 48x48 | Various places: :+1: 32x32 |
| PC | Windows XP–8.1 | Firefox 32 | :+1: 16x16 | :+1: 16x16 | :-1: 16x16 (expected 48x48) |  |
| Nexus 5 | Android 4.4.4 | Chrome 37 | :-1: 192x192 (expected 32x32) | :-1: apple-touch-icon (expected favicon) | :+1: 192x192 | Speed Dial: :-1: apple-touch-icon (expected favicon) |
| Nexus 5 | Android 4.4.4 | Firefox 32 | 48x48 (expected 32x32, but close enough) | :-1: 48x48 (expected higher) | :-1: 48x48 (expected 192x192) |  |
| Nexus 5 | Android 4.4.4 | Opera 24 | N/A | N/A | N/A | Speed Dial: :-1: apple-touch-icon (expected favicon) |
| Nexus 5 | Android 4.4.4 | Dolphin 11.2.3 | :-1: random: either 16x16 or apple-touch-icon (expected 32x32) | Can't find a view for bookmarks | :-1: apple-touch-icon (expected 192x192) | Speed Dial: :-1: apple-touch-icon (expected favicon) |
| iPhone 5s | iOS 8 | Safari 8 | N/A | N/A | :+1: apple-touch-icon |  |
| iPhone 5s | iOS 8 | Chrome 38 | :+1: 32x32 | :+1: apple-touch-icon | N/A | Speed Dial: :+1: 32x32 |
| iPhone 5s | iOS 8 | Opera Coast 3.15 | N/A | :+1: ms-tile | N/A | Speed Dial: :+1: ms-tile |

**Edit**: There is somewhat of a regression in Firefox on Android. With the RFG, Firefox always uses the 192x192 PNG, while with this method, it uses only the .ico file. While it's actually better for when a smaller icon is needed (tab, bookmark, etc.), it fails to see the large 256x256 variant for use on the home screen, where a higher resolution icon is preferred.

Unless we find something better, or any horrible problems with this approach (e.g. some other platforms don't work with this, the performance hit really is something to be concerned about), this is probably going to be my approach, as it seems to be supported best by all desktop browsers without any effect on mobile devices. I did test my iPhone and Windows tiles as well, but as I didn't change anything with the apple-touch-icons or the ms-tile icons, as expected, nothing changed on those devices.

**Comment by phbernard** [comment]
Wow! This is what I call a comprehensive feedback!

I answered your <a href="https://github.com/RealFaviconGenerator/realfavicongenerator/issues/90#issuecomment-58126728">previous comment</a> a few days ago... but apparently I didn't click the "Comment" button. Excellent. So here it is below. I'll answer your most recent comments tomorrow.

Regarding the compatibility:
- You are right, I didn't realize the 192x192 was a poor choice for most browsers. I added comments about this. I'm afraid the options are limited:
  - Remove the high resolution PNG icons. So no support for Android Chrome. Hum...
  - Report the issue to MS/IE and Apple/Safari. I already filled commented the equivalent <a href="https://bugzilla.mozilla.org/show_bug.cgi?id=751712">issue for FF</a>. I doubt this will be fixed any soon...
- <s>I got your tests, thank you. I added your results, but before I publish them, could you tell me more about the home screen icon and its blank page, default icon? Do you know how it is supposed to work? Does Safari expect any special icon? Do you know web sites that give the expected behavior?</s> You did provide this a few days later.
- The compressed icon: yep, there is a difference regarding the gradient, because the compression is lossy. This is not a problem: RFG displays a preview of the compressed icon so you can decide of the quality icon per icon. The goal of the test is to spot specific issues on some platforms... such as Mac OS (see the warning on the homepage of TinyPNG).

**Comment by phbernard** [comment]
This all sounds **very** promising. Yet, the 256x256 story is annoying: it makes the `favicon.ico` much bigger.

I did the experiment with the <a href="realfavicongenerator.net/demo_favicon.png">demo picture of RFG</a>. Here are the various sizes:
- 16x16 PNG: 1.04KB
- 32x32 PNG: 1.57KB
- 48x48 PNG: 1.71KB
- 256x256 PNG: 8.52KB
- 16x16+32x32+48x48+256x256 ICO:
  - Expected: 13KB
  - icotool: 278KB!! (this is the tool currently used by RFG. What an excellent choice).
  - ImageMagick: 23.2KB

I entered this icotool/IM thing as issue #93. By the way I already encountered the issue and <a href="http://stackoverflow.com/questions/22561648/multifile-favicon-ico-much-bigger-than-sum-of-sizes-of-component-files">there is a discussion about this on SO</a>. Without any definitive answer.

I'm going to prepare a new version of the pics, based on your findings. I'll also investigate the optimum size for the high def pic in `favicon.ico`.

**Comment by alekseyg** [comment]
> Wow! This is what I call a comprehensive feedback!

Thanks! I can get a little obsessive sometimes if something bothers me. :)

> This all sounds **very** promising. Yet, the 256x256 story is annoying: it makes the `favicon.ico` much bigger. ... I'm going to prepare a new version of the pics, based on your findings. I'll also investigate the optimum size for the high def pic in `favicon.ico`.

Sweet! I'm glad my findings were helpful. I hope you find a way to compile `favicon.ico` correctly. RFG has been a major time saver for me and I'd like to see it improve.

**Comment by phbernard** [comment]
FYI: I merged the <a href="http://realfavicongenerator.net/favicon_compatibility">results of the compatibility tests</a> you run on Mac and Android.

**Comment by smaddock** [comment]
Thanks for this @alekseyg. RFG just saved me a ton of time generating images, and your approach got the correct icons showing up in our Mac browsers.

**Comment by alekseyg** [comment]
You're welcome @smaddock
